std/machines/memory16.asm

```diff
@@ -51,26 +55,63 @@ machine Memory16(byte2: Byte2) with
     (1 - m_is_write') * (1 - m_change) * (m_value1' - m_value1) = 0;
     (1 - m_is_write') * (1 - m_change) * (m_value2' - m_value2) = 0;
 
-    col witness m_diff_lower;
-    col witness m_diff_upper;
-
     col fixed FIRST = [1] + [0]*;
     let LAST = FIRST';
 
-    link => byte2.check(m_diff_lower);
-    link => byte2.check(m_diff_upper);
-
     std::utils::force_bool(m_change);
 
     // if change is zero, addr has to stay the same.
-    (m_addr' - m_addr) * (1 - m_change) = 0;
+    (m_addr_low' - m_addr_low) * (1 - m_change) = 0;
+    (m_addr_high' - m_addr_high) * (1 - m_change) = 0;
 
-    // Except for the last row, if change is 1, then addr has to increase,
+    // Except for the last row, if m_change is 1, then addr has to increase,
     // if it is zero, step has to increase.
-    // `m_diff_upper * 2**16 + m_diff_lower` has to be equal to the difference **minus one**.
-    // Since we know that both addr and step can only be 32-Bit, this enforces that
-    // the values are strictly increasing.
-    // TODO check this
-    col diff = (m_change * (m_addr' - m_addr)) + (1 - m_change) * (m_step' - m_step);
-    (1 - LAST) * (diff - 1 - m_diff_upper * 2**16 - m_diff_lower) = 0;
+    // The diff has to be equal to the difference **minus one**.
+
+    // These two helper columns have different semantics, depending on
+    // whether we're comparing addresses or time steps.
+    // In both cases, m_tmp2 needs to be of 16 Bits.
+    col witness m_tmp1, m_tmp2;
+    link => byte2.check(m_diff);
+
+    // When comparing time steps, a 28-Bit diff is sufficient assuming a maximum step
+    // of 2**28.
+    // The difference is computed on the field, which is larger than 2**28.
+    // We prove that m_step' - m_step > 0 by letting the prover provide a 28-Bit value
+    // such that claimed_diff + 1 == m_step' - m_step.
+    // TODO: This does assume that m_step is a 28-Bit value. For rows that come from the
+    //       permutation (i.e., have a matching LHS in the main VM), this is guaranteed,
+    //       because the value comes from a fixed column. However, the prover can
+    //       "insert" reads without a matching LHS at any time (setting the selector to 0).
+    //       This allows the prover to increase the step by 2**28 in each row, over-flowing
+    //       eventually.
```

Comment:
Hmm. Maybe we to store the step in two columns as well to prove that it's 24-Bit...

Reply:
Done